mlir/lib/Gradient/Transforms/ConversionPatterns.cpp

```diff
@@ -375,6 +493,162 @@ struct BackpropOpPattern : public ConvertOpToLLVMPattern<BackpropOp> {
         return bufferSizeBytes;
     }
 
+    void convertCustomGradArgumentTypes(TypeRange argTypes,
+                                        SmallVectorImpl<Type> &llvmArgTypes) const
+    {
+        for (Type argType : argTypes) {
+            llvmArgTypes.append(2, argType);
+        }
+    }
+
+    func::FuncOp genAugmentedForward(func::FuncOp qnode, OpBuilder &builder) const
+    {
+        std::string augmentedName = (qnode.getName() + ".augfwd").str();
+        auto augmentedForward = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(
+            qnode, builder.getStringAttr(augmentedName));
+        if (augmentedForward) {
+            return augmentedForward;
+        }
+        assert(qnode.getNumResults() == 0 && "Expected QNode to be in destination-passing style");
+        MLIRContext *ctx = builder.getContext();
+        OpBuilder::InsertionGuard insertionGuard(builder);
+        builder.setInsertionPointAfter(qnode);
+        // The tape type is a null pointer because we don't need to pass any data from the forward
+        // pass to the reverse pass.
+        auto tapeType = LLVM::LLVMPointerType::get(ctx);
+        SmallVector<Type> argTypes;
+        convertCustomGradArgumentTypes(qnode.getArgumentTypes(), argTypes);
+        augmentedForward = builder.create<func::FuncOp>(
+            qnode.getLoc(), augmentedName, FunctionType::get(ctx, argTypes, {tapeType}));
+        augmentedForward.setPrivate();
+        Location loc = qnode.getLoc();
+
+        Block *entry = augmentedForward.addEntryBlock();
+        builder.setInsertionPointToStart(entry);
+
+        // Every other argument is a shadow
+        SmallVector<Value> arguments;
+        for (unsigned i = 0; i < qnode.getNumArguments(); i++) {
+            arguments.push_back(augmentedForward.getArgument(i * 2));
+        }
+
+        builder.create<func::CallOp>(loc, qnode, arguments);
```

Comment:
From what I understand we just have a simple forward pass?

Reply:
Yes, the augmented forward pass does the exact same thing as the regular forward pass. The reason is that Enzyme might need the result of the regular forward pass.

If you have postprocessing like `q = quantum_circuit(...); return q ** 2`, at some point Enzyme will need to compute `2 * q` as part of its derivative computation.

This will currently result in at least two forward passes being computed (one in the augmented forward pass, another in the `qgradFn`). We could potentially optimize this in the future to avoid the second forward pass, especially when differentiating with the adjoint method.

Reply:
Ok good to know, I think it is already in the document?

Reply:
Yup, I've added the various clarification points from our review discussion to [the notion document](https://www.notion.so/xanaduai/Enzyme-Driven-Hybrid-Gradients-19577a4d03964b279c5c3872c4d6d619).